### PR TITLE
Display favourite messages in a room-like view

### PR DESCRIPTION
**Pull request by andybalaam**
Originally developed by @yaya-usman but remodelled quite a bit now by @andybalaam 

This creates a new view that you can access in a similar way to a room:
image

It displays all your favourite messages.

Apologies for the huge review. Do ask me and I can walk you through it interactively.

---
Here's what your changelog entry will look like:

## ✨ Features
 * Display favourite messages in a room-like view ([\#9719](https://github.com/matrix-org/matrix-react-sdk/pull/9719)). Contributed by @andybalaam.

### CODE REVIEW

**Comment by andybalaam** [comment]
Work will continue on this in January

**Comment by andybalaam** [comment]
Prerequisite: https://github.com/matrix-org/matrix-analytics-events/pull/76

File: src/components/structures/FavouriteMessagesView/ConfirmClearDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+import React, { FC, useCallback } from "react";
+
+import useFavouriteMessages from "../../../hooks/useFavouriteMessages";
+import { _t } from "../../../languageHandler";
+import BaseDialog from "../../views/dialogs/BaseDialog";
+import { IDialogProps } from "../../views/dialogs/IDialogProps";
+import DialogButtons from "../../views/elements/DialogButtons";
+
+/*
```
Comment (turt2live) [inline src/components/structures/FavouriteMessagesView/ConfirmClearDialog.tsx:25]:
```suggestion
/**
```

File: src/components/structures/FavouriteMessagesView/FavouriteMessagesPanel.tsx [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                <FavouriteMessagesHeader query={props.searchQuery} handleSearchQuery={props.handleSearchQuery} />
+                <h2 className="mx_FavouriteMessages_emptyMarker">{_t("No Favourite Messages")}</h2>
+            </>
+        );
+    } else {
+        return (
+            <>
+                <FavouriteMessagesHeader query={props.searchQuery} handleSearchQuery={props.handleSearchQuery} />
+                <ScrollPanel
+                    ref={favouriteMessagesPanelRef}
```
Comment (turt2live) [inline src/components/structures/FavouriteMessagesView/FavouriteMessagesPanel.tsx:49]: looks like a legitimate strict error, albeit annoying: 
```suggestion
                    ref={favouriteMessagesPanelRef ?? null}
```

File: src/components/structures/FavouriteMessagesView/FavouriteMessagesTilesList.tsx [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+                    <li key={mxEvent.getId() + "-room"}>
+                        <h2>
+                            {_t("Room")}: {room ? room.name : ""}
+                        </h2>
+                    </li>,
+                );
+                lastRoomId = roomId!;
+            }
+
+            const resultLink = "#/room/" + roomId + "/" + mxEvent.getId();
```
Comment (turt2live) [inline src/components/structures/FavouriteMessagesView/FavouriteMessagesTilesList.tsx:74]: this appears to be an assumption of our url routing which isn't guaranteed - can we convert it to a `view_room` dispatch instead?

File: src/components/structures/FavouriteMessagesView/FavouriteMessagesView.tsx [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+ */
+async function calcEvents(
+    searchQuery: string,
+    favouriteMessages: FavouriteStorage[],
+    matrixClient: MatrixClient,
+): Promise<MatrixEvent[]> {
+    const displayedFavourites: FavouriteStorage[] = filterFavourites(searchQuery, favouriteMessages);
+    const promises: Promise<MatrixEvent | null>[] = displayedFavourites.map((f) => fetchEvent(f, matrixClient));
+    const events = await Promise.all(promises);
+    return events.filter((e) => e !== null);
```
Comment (turt2live) [inline src/components/structures/FavouriteMessagesView/FavouriteMessagesView.tsx:131]:
```suggestion
    return events.filter((e) => e !== null) as MatrixEvent[]; // force cast because typescript doesn't understand what `filter` does
```

File: src/components/views/messages/MessageActionBar.tsx [hunk 5]
```diff
@@ -442,7 +445,9 @@ export default class MessageActionBar extends React.PureComponent<IMessageAction
 
     public render(): JSX.Element {
         const toolbarOpts = [];
-        if (canEditContent(this.props.mxEvent)) {
+        const isFavouritePageCheck = window.location.href.split("/").pop() === "favourite_messages";
```
Comment (turt2live) [inline src/components/views/messages/MessageActionBar.tsx:448]: we shouldn't be making assumptions about the URL structure, or whether it's at the root level. We should be able to ask MatrixChat or similar if we're on the page, though.

File: src/dispatcher/actions.ts [hunk 6]
```diff
@@ -346,4 +346,14 @@ export enum Action {
...
+
+    /**
+     * Fired when we want to view favourited messages panel
+     */
+    ViewFavouriteMessages = "view_favourite_messages",
+
+    /**
+     * Fired when we want to clear all favourited messages
+     */
+    OpenClearModal = "open_clear_modal",
```
Comment (turt2live) [inline src/dispatcher/actions.ts:358]: this should be named with favourited messages in mind - the documentation is clear, but the enum is not self-documenting.

File: src/dispatcher/actions.ts [hunk 7]
```diff
@@ -346,4 +346,14 @@ export enum Action {
      * Fired when we want to view a thread, either a new one or an existing one
      */
     ShowThread = "show_thread",
+
+    /**
+     * Fired when we want to view favourited messages panel
```
Comment (turt2live) [inline src/dispatcher/actions.ts:351]:
```suggestion
     * Fired when we want to view favourited messages panel. No additional payload information required.
```

File: src/dispatcher/actions.ts [hunk 8]
```diff
@@ -346,4 +346,14 @@ export enum Action {
...
+     */
+    ViewFavouriteMessages = "view_favourite_messages",
+
+    /**
+     * Fired when we want to clear all favourited messages
```
Comment (turt2live) [inline src/dispatcher/actions.ts:356]:
```suggestion
     * Fired when we want to clear all favourited messages. No additional payload information required.
```

File: src/hooks/useFavouriteMessages.ts [hunk 9]
```diff
@@ -14,27 +14,114 @@ See the License for the specific language governing permissions and
...
+    roomId: string;
+    content: IContent;
+}
+
+// Global variable tracking LocalStorage state
+let ioElementFavouriteMessages: FavouriteStorage[] | null = null;
+
+function loadFavourites(): FavouriteStorage[] {
+    try {
+        return JSON.parse(window.localStorage?.getItem("io_element_favouriteMessages") ?? "[]");
```
Comment (turt2live) [inline src/hooks/useFavouriteMessages.ts:31]: This should be run through the SettingsStore like breadcrumbs, and because we're in the react-sdk be prefixed with `mx` rather than Element.

File: package.json [hunk 10]
```diff
@@ -56,7 +56,7 @@
     },
     "dependencies": {
         "@babel/runtime": "^7.12.5",
-        "@matrix-org/analytics-events": "^0.3.0",
+        "@matrix-org/analytics-events": "^0.4.0",
```
Comment (turt2live) [inline package.json:59]: this change appears unrelated

**Review by turt2live — CHANGES_REQUESTED** [review]
generally seems good, thank you!

File: package.json (hunk 10 above)
Comment (andybalaam) [reply package.json:59]: It's needed because I added Favourite Messages to WebScreen in that version (https://github.com/matrix-org/matrix-analytics-events/commit/4ef88563c8dc3f5a5bd8ed1194511ab597c98b7b).

Comment (turt2live) [reply package.json:59]: That constant doesn't appear to be used though? Or I'm not seeing the magic that glues screens to analytics-events.

File: src/hooks/useFavouriteMessages.ts (hunk 9 above)
Comment (andybalaam) [reply src/hooks/useFavouriteMessages.ts]: local storage key changed in 7f934326af2

Comment (andybalaam) [reply src/hooks/useFavouriteMessages.ts]: > This should be run through the SettingsStore like breadcrumbs

Not sure what you mean - this is not where we are storing the on/off setting - this is where we are storing the actual information about what is favourited. Do you think that should be inside SettingsStore?

Comment (t3chguy) [reply src/hooks/useFavouriteMessages.ts]: image

These are all similar to that, using SettingsStore just means you get the abstraction to be able to use matrix account data storage for free

File: src/dispatcher/actions.ts (hunk 6 above)
Comment (andybalaam) [reply src/dispatcher/actions.ts]: Fixed in e64308aa3bc45

File: src/hooks/useFavouriteMessages.ts (hunk 9 above)
Comment (andybalaam) [reply src/hooks/useFavouriteMessages.ts]: Wow, cool! Thanks!

File: src/components/structures/FavouriteMessagesView/FavouriteMessagesTilesList.tsx (hunk 3 above)
Comment (andybalaam) [reply src/components/structures/FavouriteMessagesView/FavouriteMessagesTilesList.tsx:74]: It's copied from here: https://github.com/matrix-org/matrix-react-sdk/blob/develop/src/components/structures/RoomSearchView.tsx#L255

I presume that could do with fixing too?

File: src/hooks/useFavouriteMessages.ts (hunk 9 above)
Comment (andybalaam) [reply src/hooks/useFavouriteMessages.ts]: Done in the version I will push soon (rebased, because it got really messy)

File: src/components/structures/FavouriteMessagesView/FavouriteMessagesTilesList.tsx (hunk 3 above)
Comment (andybalaam) [reply src/components/structures/FavouriteMessagesView/FavouriteMessagesTilesList.tsx:74]: Looking into the code, it seems that `EventTile` does not currently support an `onClick` or similar event, so the only way to allow clicking a tile to jump somewhere else is with its `highlightLink` property as we have done here and in `RoomSearchView`.

Given how long this PR has festered, I'd like to argue that fixing this is outside of the scope of this PR, and something we can do later.

**Comment by andybalaam** [comment]
i18n test is failing with

>   fatal: couldn't find remote ref andybalaam/favourite-messages-rebased

which I don't understand

**Comment by andybalaam** [comment]
OK, CI seems happy so requesting a re-review from @t3chguy

**Comment by andybalaam** [comment]
> OK, CI seems happy so requesting a re-review from @t3chguy

Scratch that - techguy is too busy today so I'll ask someone else.

File: res/css/structures/_FavouriteMessagesView.pcss [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    mask-image: url("$(res)/img/element-icons/room/sort-twoway.svg");
+}
+
+.mx_FavMessagesHeader_clearAllButton::before {
+    mask-image: url("$(res)/img/element-icons/room/clear-all.svg");
+}
+
+.mx_FavMessagesHeader_cancelButton {
+    background-color: $alert;
+    mask: url("$(res)/img/cancel.svg");
```
Comment (florianduros) [inline res/css/structures/_FavouriteMessagesView.pcss:69]: We are trying to get rid of mask in favour of the `svg` as React component directly

Example of use:
https://github.com/matrix-org/matrix-react-sdk/blob/develop/src/components/views/rooms/wysiwyg_composer/components/FormattingButtons.tsx#L21

File: res/css/structures/_FavouriteMessagesView.pcss [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    .mx_FavMessagesHeader_Wrapper_right {
+        display: flex;
+        align-items: center;
+        flex: 0.6;
+        justify-content: flex-end;
+    }
+}
+
+.mx_FavMessagesHeader_sortButton::before {
+    mask-image: url("$(res)/img/element-icons/room/sort-twoway.svg");
```
Comment (florianduros) [inline res/css/structures/_FavouriteMessagesView.pcss:60]: Same comment about the mask

File: res/css/structures/_FavouriteMessagesView.pcss [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+.mx_FavMessagesHeader_clearAllButton::before {
+    mask-image: url("$(res)/img/element-icons/room/clear-all.svg");
```
Comment (florianduros) [inline res/css/structures/_FavouriteMessagesView.pcss:64]: Same comment about the mask

File: res/css/structures/_FavouriteMessagesView.pcss [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+    .mx_FavMessagesHeader_Wrapper_left {
+        display: flex;
+        align-items: center;
+        flex: 0.4;
+
+        & > span {
+            color: $primary-content;
+            font-weight: $font-semi-bold;
+            font-size: $font-18px;
+            margin: 0 8px;
```
Comment (florianduros) [inline res/css/structures/_FavouriteMessagesView.pcss:47]: We can use here `$spacing-8`

**Comment by weeman1337** [comment]
Played a bit around with it on the deployment here. I like it :+1: 
Some issues I found:

If you favourite a poll, it explodes with:

```
Uncaught (in promise) TypeError: f.content.body is undefined
    filterFavourites FavouriteMessagesView.tsx:72
    filterFavourites FavouriteMessagesView.tsx:71
    calcEvents FavouriteMessagesView.tsx:134
    recalcEvents FavouriteMessagesView.tsx:43
    FavouriteMessagesView FavouriteMessagesView.tsx:49
    fk React
    unstable_runWithPriority scheduler.production.min.js:18
    React 4
    unstable_runWithPriority scheduler.production.min.js:18
    React 6
    setState MatrixChat.tsx:414
    setPage MatrixChat.tsx:888
    viewFavouriteMessages MatrixChat.tsx:1051
    MatrixChat MatrixChat.tsx:848
    _invokeCallback Dispatcher.js:214
    dispatch Dispatcher.js:189
    setTimeout handler*dispatch dispatcher.ts:52
    onFavouriteClicked RoomList.tsx:716
    React 11
    unstable_runWithPriority scheduler.production.min.js:18
    React 11
    loadApp init.tsx:150
    start index.ts:228
    async* index.ts:239
    Webpack 3
[FavouriteMessagesView.tsx:72:9](webpack:///src/components/structures/FavouriteMessagesView/FavouriteMessagesView.tsx)
```

(same for broadcasts)

---

You cannot jump to the room message, if you favourite an image, voice message or location share (click the message in the favourites view).

---

The room separation in the favourite view could be improved.

image

Maybe just add more spacing above the room names.

---

If you favourite a thread root, the thread summary is shown in favourites. But clicking it does nothing.

File: res/css/structures/_FavouriteMessagesView.pcss [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+.mx_FavMessagesHeader_cancelButton {
+    background-color: $alert;
+    mask: url("$(res)/img/cancel.svg");
+    mask-repeat: no-repeat;
+    mask-position: center;
+    mask-size: 17px;
+    padding: 9px;
+    margin: 0 12px 0 3px;
```
Comment (florianduros) [inline res/css/structures/_FavouriteMessagesView.pcss:74]: Use `$spacing-*`

File: res/css/structures/_FavouriteMessagesView.pcss [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+    z-index: 999;
+}
+
+.mx_FavMessagesHeader_Wrapper {
+    height: 44px;
+    display: flex;
+    align-items: center;
+    min-width: 0;
+    margin: 0 20px 0 16px;
+    padding-top: 8px;
```
Comment (florianduros) [inline res/css/structures/_FavouriteMessagesView.pcss:34]: Use `$spacing-*`

File: res/css/structures/_FavouriteMessagesView.pcss [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+    background-color: $background;
+    z-index: 999;
+}
+
+.mx_FavMessagesHeader_Wrapper {
+    height: 44px;
+    display: flex;
+    align-items: center;
+    min-width: 0;
+    margin: 0 20px 0 16px;
```
Comment (florianduros) [inline res/css/structures/_FavouriteMessagesView.pcss:33]: Use `$spacing-*`

**Comment by bwindels** [comment]
If I read correctly, this will save all favorite message ids in a single account data entry (`favourite_messages` with element settings prefix). The downsides of this approach would be that the array could grow quite large, and also data loss in case of a race between multiple clients/devices (a pain currently with the DM rooms mapping). Curious what your thoughts are on these, and if storing a single favorite (or only favorites for one room) per account data entry was considered. I can see that would probably be harder to integrate with the element settings code though.
The reason I bring it up (of lesser importance) is also that a pattern like `io.element.favourite_messages.<roomid>.<eventid>` for the type with only 1 favourite per account data entry would make things easier in Hydrogen as well to only load in memory what is needed.

File: res/css/structures/_FavouriteMessagesView.pcss [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+
+.mx_FavMessagesHeader {
+    position: fixed;
+    top: 0;
+    left: 0;
+    width: 100%;
+    flex: 0 0 50px;
+    border-bottom: 1px solid $primary-hairline-color;
+    background-color: $background;
+    z-index: 999;
```
Comment (florianduros) [inline res/css/structures/_FavouriteMessagesView.pcss:25]: Is that necessary ? 😅

File: res/img/element-icons/room/sort-twoway.svg [hunk 19]
```diff
@@ -0,0 +1,11 @@
+<svg xmlns="http://www.w3.org/2000/svg" x="0px" y="0px"
+	 width="24px" height="24px" viewBox="0 0 52 52" xml:space="preserve">
```
Comment (florianduros) [inline res/img/element-icons/room/sort-twoway.svg:2]: `fill="currentColor"` is not needed ?

File: src/components/structures/FavouriteMessagesView/FavouriteMessageTile.tsx [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+
+const FavouriteMessageTile: FC<IProps> = (props: IProps) => {
+    let context!: React.ContextType<typeof RoomContext>;
+
+    const result = props.result;
+    const eventId = result.getId();
+
+    const ts1 = result?.getTs();
+    const ret = [<DateSeparator key={ts1 + "-search"} roomId={result.getRoomId()!} ts={ts1} />];
+    const layout = SettingsStore.getValue<Layout>("layout");
```
Comment (florianduros) [inline src/components/structures/FavouriteMessagesView/FavouriteMessageTile.tsx:56]: Maybe, we can use `useSettingValue` here. It will allow to have a live reload when the value in the settings is changed. It will trigger a rerender of the component

**Comment by andybalaam** [comment]
> If I read correctly, this will save all favorite message ids in a single account data entry (`favourite_messages` with element settings prefix). The downsides of this approach would be that the array could grow quite large, and also data loss in case of a race between multiple clients/devices (a pain currently with the DM rooms mapping). Curious what your thoughts are on these, and if storing a single favorite (or only favorites for one room) per account data entry was considered. I can see that would probably be harder to integrate with the element settings code though. The reason I bring it up (of lesser importance) is also that a pattern like `io.element.favourite_messages.<roomid>.<eventid>` for the type with only 1 favourite per account data entry would make things easier in Hydrogen as well to only load in memory what is needed.

Because this PR started out using local storage, I totally forgot when I converted it to use SettingsStorage that there is an MSC we are planning to follow: https://github.com/matrix-org/matrix-spec-proposals/pull/2437

Note: that MSC proposes using Room Account Data, but I plan to use general Account Data because I think favourites should not be tied to rooms.

@bwindels I think that MSC also uses a single account data entry for everything - will that cause you similar problems?

File: src/components/structures/FavouriteMessagesView/FavouriteMessageTile.tsx [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+            const prevEv = props.timeline[j - 1];
+            // is this a continuation of the previous message?
+            const continuation =
+                !dateSeparator(prevEv, mxEv) &&
+                shouldFormContinuation(prevEv, mxEv, context?.showHiddenEvents, threadsEnabled);
+
+            let lastInSection = true;
+            const nextEv = props?.timeline[j + 1];
+            if (nextEv) {
+                lastInSection =
```
Comment (florianduros) [inline src/components/structures/FavouriteMessagesView/FavouriteMessageTile.tsx:76]:
```js
const lastInSection = nextEv && 
                 ( dateSeparator(mxEv, nextEv) ||
                    mxEv.getSender() !== nextEv.getSender() ||
                    !shouldFormContinuation(mxEv, nextEv, context?.showHiddenEvents, threadsEnabled)
                    )
```

**Comment by bwindels** [comment]
> > If I read correctly, this will save all favorite message ids in a single account data entry (`favourite_messages` with element settings prefix). The downsides of this approach would be that the array could grow quite large, and also data loss in case of a race between multiple clients/devices (a pain currently with the DM rooms mapping). Curious what your thoughts are on these, and if storing a single favorite (or only favorites for one room) per account data entry was considered. I can see that would probably be harder to integrate with the element settings code though. The reason I bring it up (of lesser importance) is also that a pattern like `io.element.favourite_messages.<roomid>.<eventid>` for the type with only 1 favourite per account data entry would make things easier in Hydrogen as well to only load in memory what is needed.
> 
> Because this PR started out using local storage, I totally forgot when I converted it to use SettingsStorage that there is an MSC we are planning to follow: [matrix-org/matrix-spec-proposals#2437](https://github.com/matrix-org/matrix-spec-proposals/pull/2437)


Thanks for the MSC link. I always forget about room account data :) The size of the array would indeed be more reasonable when using room account data. The race is indeed describes as a [limitation](Thanks for the MSC link. I always forget about room account data :) The size of the array would indeed be more reasonable when using room account data), which is a long standing issue with account data, and the proposed solution has been proposed before I think (using ETag and If-Match to detect conflicts), and would indeed be a good fix but hasn't been implemented so far. So yeah, a discussion for the MSC. Perhaps it's fine to have an eventual solution planned for now without needing to implement it?

> 
> Note: that MSC proposes using Room Account Data, but I plan to use general Account Data because I think favourites should not be tied to rooms.

But favourites are always for events, which can only occur within a room, no?

> 
> @bwindels I think that MSC also uses a single account data entry for everything - will that cause you similar problems?

File: src/components/structures/FavouriteMessagesView/FavouriteMessagesView.tsx [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+        const newEvents = await calcEvents(searchQuery, faves, matrixClient);
+        setFavouriteMessageEvents(newEvents);
+    }, [searchQuery, matrixClient, getFavouriteMessages]);
+
+    // Because finding events is async, we do it in useEffect, not useState.
+    useEffect(() => {
+        recalcEvents();
+    }, [searchQuery, recalcEvents]);
+
+    registerFavouritesChangedListener(() => recalcEvents());
```
Comment (florianduros) [inline src/components/structures/FavouriteMessagesView/FavouriteMessagesView.tsx:52]: At every render, it will call the `registerFavouritesChangedListener`  and a listener

File: src/components/structures/FavouriteMessagesView/FavouriteMessagesView.tsx [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+ */
+async function calcEvents(
+    searchQuery: string,
+    favouriteMessages: FavouriteStorage[],
+    matrixClient: MatrixClient,
+): Promise<MatrixEvent[]> {
+    const displayedFavourites: FavouriteStorage[] = filterFavourites(searchQuery, favouriteMessages);
+    const promises: Promise<MatrixEvent | null>[] = displayedFavourites.map((f) => fetchEvent(f, matrixClient));
+    const events = await Promise.all(promises);
+    return events.filter((e) => e !== null) as MatrixEvent[]; // force cast because typescript doesn't understand what `filter` does
```
Comment (florianduros) [inline src/components/structures/FavouriteMessagesView/FavouriteMessagesView.tsx:137]: We can use https://stackoverflow.com/questions/43118692/typescript-filter-out-nulls-from-an-array this kind of predicate to cleanly filter without casting the type

File: src/components/views/messages/MessageActionBar.tsx [hunk 24]
```diff
@@ -274,14 +275,19 @@ const ReplyInThreadButton: React.FC<IReplyInThreadButton> = ({ mxEvent }) => {
...
 }
 
-const FavouriteButton: React.FC<IFavouriteButtonProp> = ({ mxEvent }) => {
-    const { isFavourite, toggleFavourite } = useFavouriteMessages();
+const FavouriteButton: FC<IFavouriteButtonProp> = ({ mxEvent, favouriteMessagesStore }: IFavouriteButtonProp) => {
+    const { isFavourite, toggleFavourite, registerFavouritesChangedListener } =
+        useFavouriteMessages(favouriteMessagesStore);
+    const [, forceRefresh] = useState([]);
+
+    registerFavouritesChangedListener(() => forceRefresh([]));
```
Comment (florianduros) [inline src/components/views/messages/MessageActionBar.tsx:286]: It seems to be a bit hacky here, maybe `useFavouriteMessages` should return something to trigger the rerender ?

File: src/hooks/useFavouriteMessages.ts [hunk 25]
```diff
@@ -14,30 +14,61 @@ See the License for the specific language governing permissions and
...
-    toggleFavourite: (eventId: string) => void;
+export default function useFavouriteMessages(favouriteMessageStore = FavouriteMessagesStore.instance): {
+    getFavouriteMessages: () => FavouriteStorage[];
     isFavourite: (eventId: string) => boolean;
+    toggleFavourite: (mxEvent: MatrixEvent) => void;
+    clearFavouriteMessages: () => void;
+    registerFavouritesChangedListener: (listener: () => void) => void;
 } {
-    const [, setX] = useState<string[]>();
+    const myListeners: (() => void)[] = [];
```
Comment (florianduros) [inline src/hooks/useFavouriteMessages.ts:29]: At every render, `myListeners` is replaced by an empty array which seems to be unwanted behaviour. 

It needs to be a `useState` or `useRef`

File: src/components/structures/FavouriteMessagesView/FavouriteMessagesView.tsx (hunk 22 above)
Comment (andybalaam) [reply src/components/structures/FavouriteMessagesView/FavouriteMessagesView.tsx:52]: Do you have an alternative suggestion?

**Comment by andybalaam** [comment]
> > Note: that MSC proposes using Room Account Data, but I plan to use general Account Data because I think favourites should not be tied to rooms.

> But favourites are always for events, which can only occur within a room, no?

Yes, but I argue that:

a) When I want to view my favourites I don't want to iterate all rooms
b) When I leave a room I don't want to lose the information that I favourited a message inside it

File: src/components/structures/FavouriteMessagesView/FavouriteMessagesView.tsx (hunk 22 above)
Comment (florianduros) [reply src/components/structures/FavouriteMessagesView/FavouriteMessagesView.tsx:52]: We can add the listener in a `useEffect`:

```
useEffect(() => {
     const handler = () => recalcEvents();
     registerFavouritesChangedListener(handler);
     return () => {
         // here we would like to unregister the handler
         unregisterFavouritesChangedListener(handler);
     }
}, [registerFavouritesChangedListener, recalcEvents]
```

If we want to avoid to repeat this behaviour every time we use the `useFavourites` hook. We can pass the handler as argument of the hook and make all the register/unregister in it.

**Review by weeman1337 — CHANGES_REQUESTED** [review]
Some minor findings in the code.

**Comment by weeman1337** [comment]
This is the last one for today. And maybe I am picky here :upside_down_face: The header design for favourite message is not the same as for rooms:

header

**Comment by andybalaam** [comment]
> This is the last one for today. And maybe I am too picky here, but it hurts me upside_down_face The header design for favourite message is not the same as for rooms:

Not too picky! Well spotted.

**Comment by andybalaam** [comment]
This is all awesome feedback - thank you @weeman1337 @florianduros @bwindels !

I was hoping to squeeze the work to get this PR into shape in between other things, and given how much work there is left to do, I don't think I can do that right now :-(

So, any of the following would be much appreciated:

* Help with fixing some of the above comments
* Suggestions for how to break this down into more manageable pieces so at least some of it can be merged

Thanks!

**Comment by andybalaam** [comment]
Not currently being worked on, so moving to draft.
